compiler/rustc_parse/src/parser/diagnostics.rs

```diff
@@ -1330,6 +1337,37 @@ impl<'a> Parser<'a> {
         }
     }
 
+    /// Rust has no ternary operator (`cond ? then : else`). Parse it and try
+    /// to recover from it if `then` and `else` are valid expressions.
+    pub(super) fn maybe_recover_from_ternary_operator(&mut self) -> PResult<'a, ()> {
+        let snapshot = self.create_snapshot_for_diagnostic();
+        let Some(lo) = self.maybe_ternary_lo else {
+            return Ok(());
+        };
+
+        if self.prev_token == token::Question {
+            if let Err(err) = self.parse_expr() {
```

Comment:
Can you explain why we're recovering the failure case here? Does `parse_expr` try to eat the `:` here? I would've expected it to leave the colon behind, i.e. for the code to look more like:

```
match self.parse_expr() {
  Ok(_) if self.token == token::Colon => { /* keep going */ },
  Ok(_) => { self.restore_snapshot(snapshot); return Ok(()); }
  Err(e) => { e.cancel(); self.restore_snapshot(snapshot); return Ok(()); }
}

self.bump(); // eat `:`
```

Reply:
Kind of, it thinks it's type ascription unfortunately on cases like `f32::MAX` and errors on it. Not sure if there's a better way to do it.

Reply:
See my comment below about possibly implementing this logic in `parse_expr_dot_or_call_with_` with eager lookahead and `Parser::may_recover` instead.

Reply:
I've implemented it there now, it still needs to do this though. It basically does the inverse of what it did before: It checks if it successfully parses an expression, then a `:`, then another expression after a `?`. Shouldn't cause any noticeable issues but checking this on *every* `?` is terrible, imo, so is there a better way to do that?

Reply:
Ok it caused noticeable issues :sweat_smile:, looks like it also checks after a match on an `expr?` which means an unexpected `if` may appear. All good, except for the fact that it emits this error without returning `Err`... So we can't cancel it.

This will likely need a different approach.

Reply:
Ugh, yeah, there are cases where this may be ambiguous, e.g. `expr ? - expr`

Reply:
That would fail to find the `: expr`, though. So I *think* it won't complain about the ternary unless there's an invalid expression, then a `:`, then an actual expression, which is never valid syntax afaik. Unless I'm misunderstanding?

Regardless though, is there a way to 'silence' diagnostics temporarily, like with snapshots?

Reply:
No you're correct, but we should probably only be doing this somewhat expensive parsing if we know we're (almost surely) on an error path, which is pretty difficult to form a heuristic for..

Reply:
It's probably easier to move it back to `expected_one_of_not_found` then, but store a mutable reference to the `?`. That way we can modify it if we're allowed to recover. That way we for sure only do it on an error path

Reply:
Ignore i.e. keep them or somehow suppress them? I'm fine with either.

Reply:
Keep them